File: WebClient SDK/src/default/ui/krProcess/stageHandlers/approvalUIHandler.ts
import { CardFieldChangedEventArgs, CardRow } from 'tessa/cards';
import { DefaultCardTypes, plainColumnName } from 'tessa/workflow';
import { DotNetType, Guid, Visibility } from 'tessa/platform';
import { IBlockViewModel, IControlViewModel } from 'tessa/ui/cards';
import {
  IKrStageTypeUIHandlerContext,
  KrStageTypeUIHandler,
  StageTypeHandlerDescriptor,
  approvalDescriptor
} from 'tessa/workflow/krProcess';
import { RequestParameterBuilder, TessaViewRequest, ViewService } from 'tessa/views';

import { TabControlViewModel } from 'tessa/ui/cards/controls';
import { equalsCriteriaOperator } from 'tessa/views/metadata';

/**
 * UI обработчик типа этапа {@link approvalDescriptor}.
 */
export class ApprovalUIHandler extends KrStageTypeUIHandler {
  //#region fields

  /**
   * Идентификатор карточки вида задания "Рекомендательное согласование".
   */
  private static readonly _advisoryTaskKindId: guid = '2e6c5d3e-d408-4f98-8a55-e9d1316bf2cc';

  private static readonly _advisoryField: string = plainColumnName(
    'KrApprovalSettingsVirtual',
    'Advisory'
  );

  private static readonly _notReturnEditField: string = plainColumnName(
    'KrApprovalSettingsVirtual',
    'NotReturnEdit'
  );

  private static readonly _kindIdField: string = plainColumnName(
    'KrTaskKindSettingsVirtual',
    'KindID'
  );

  private static readonly _kindCaptionField: string = plainColumnName(
    'KrTaskKindSettingsVirtual',
    'KindCaption'
  );

  private static readonly _returnWhenDisapprovedField: string = plainColumnName(
    'KrApprovalSettingsVirtual',
    'ReturnWhenDisapproved'
  );

  private _settings?: CardRow;
  private _returnIfNotApprovedFlagControl?: IControlViewModel;
  private _returnAfterApprovalFlagControl?: IControlViewModel;

  //#endregion

  //#region base overrides

  public descriptors(): StageTypeHandlerDescriptor[] {
    return [approvalDescriptor];
  }

  public async initialize(context: IKrStageTypeUIHandlerContext): Promise<void> {
    let flagsBlock: IBlockViewModel | undefined;
    let flagsTabs: IControlViewModel | undefined;

    if (
      (flagsBlock = context.settingsForms
        .find(x => x.name === DefaultCardTypes.KrApprovalStageTypeSettingsTypeName)
        ?.blocks.find(x => x.name === 'ApprovalStageFlags')) &&
      (flagsTabs = flagsBlock.controls.find(x => x.name === 'FlagsTabs')) &&
      flagsTabs instanceof TabControlViewModel
    ) {
      this._returnIfNotApprovedFlagControl = flagsTabs.tabs
        .find(x => x.name === 'CommonSettings')
        ?.blocks.find(x => x.name === 'StageFlags')
        ?.controls.find(x => x.name === 'ReturnIfNotApproved');

      this._returnAfterApprovalFlagControl = flagsTabs.tabs
        .find(x => x.name === 'AdditionalSettings')
        ?.blocks.find(x => x.name === 'StageFlags')
        ?.controls.find(x => x.name === 'ReturnAfterApproval');
    }

    this._settings = context.row;
    this._settings.fieldChanged.add(this.onSettingsFieldChanged);

    this.advisoryConfigureFields(this._settings.tryGet(ApprovalUIHandler._advisoryField));
    this.notReturnEditConfigureFields(this._settings.tryGet(ApprovalUIHandler._notReturnEditField));
  }

  public async finalize(): Promise<void> {
    if (this._settings) {
      this._settings.fieldChanged.remove(this.onSettingsFieldChanged);
      this._settings = undefined;
    }
  }

  //#endregion

  //#region private methods

  private readonly onSettingsFieldChanged = async (e: CardFieldChangedEventArgs): Promise<void> => {
    if (e.fieldName === ApprovalUIHandler._advisoryField) {
      const advisory: boolean = e.fieldValue;

      this.advisoryConfigureFields(advisory);

      if (advisory) {
        if (!this._settings!.tryGet(ApprovalUIHandler._kindIdField)) {
          const { kindId, kindCaption } = await this.getKind(ApprovalUIHandler._advisoryTaskKindId);

          if (kindId) {
            this._settings!.set(ApprovalUIHandler._kindIdField, kindId, DotNetType.Guid);
            this._settings!.set(
              ApprovalUIHandler._kindCaptionField,
              kindCaption,
              DotNetType.String
            );
          }
        }
      } else {
        if (
          Guid.equals(
            this._settings!.tryGet(ApprovalUIHandler._kindIdField),
            ApprovalUIHandler._advisoryTaskKindId
          )
        ) {
          this._settings!.set(ApprovalUIHandler._kindIdField, null);
          this._settings!.set(ApprovalUIHandler._kindCaptionField, null);
        }

        if (this._returnIfNotApprovedFlagControl) {
          this._returnIfNotApprovedFlagControl.isReadOnly = false;
        }
      }

      return;
    }

    if (e.fieldName === ApprovalUIHandler._notReturnEditField) {
      this.notReturnEditConfigureFields(e.fieldValue);
    }
  };

  private async getKind(id: guid): Promise<{ kindId: guid | null; kindCaption: string }> {
    const taskKindsView = ViewService.instance.getByName('TaskKinds')!;

    const request = new TessaViewRequest(taskKindsView.metadata);

    const idParam = new RequestParameterBuilder()
      .withMetadata(taskKindsView.metadata.parameters.get('ID')!)
      .addCriteria(equalsCriteriaOperator(), '', id)
      .asRequestParameter();

    request.values.push(idParam);

    const result = await taskKindsView.getData(request);

    if (!result.rows || result.rows.length === 0) {
      return { kindId: null, kindCaption: '' };
    }

    const row = result.rows[0];
    return { kindId: row[0], kindCaption: row[1] };
  }

  private advisoryConfigureFields(isAdvisory: boolean) {
    if (isAdvisory) {
      if (this._returnIfNotApprovedFlagControl) {
        this._returnIfNotApprovedFlagControl.isReadOnly = true;
        this._settings!.set(
          ApprovalUIHandler._returnWhenDisapprovedField,
          false,
          DotNetType.Boolean
        );
      }
    }
  }

  private notReturnEditConfigureFields(isNotReturnEdit: boolean) {
    if (isNotReturnEdit) {
      if (this._returnIfNotApprovedFlagControl) {
        this._returnIfNotApprovedFlagControl.controlVisibility = Visibility.Collapsed;
      }

      if (this._returnAfterApprovalFlagControl) {
        this._returnAfterApprovalFlagControl.controlVisibility = Visibility.Collapsed;
      }
    } else {
      if (this._returnIfNotApprovedFlagControl) {
        this._returnIfNotApprovedFlagControl.controlVisibility = Visibility.Visible;
      }

      if (this._returnAfterApprovalFlagControl) {
        this._returnAfterApprovalFlagControl.controlVisibility = Visibility.Visible;
      }
    }
  }

  //#endregion
}

File: Source/Extensions/Tessa.Extensions.Default.Client/UI/KrProcess/StageHandlers/ApprovalUIHandler.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform;
using Tessa.Platform.Storage;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;
using Tessa.Views;
using Tessa.Views.Metadata.Criteria;

namespace Tessa.Extensions.Default.Client.UI.KrProcess.StageHandlers
{
    /// <summary>
    /// UI обработчик типа этапа <see cref="StageTypeDescriptors.ApprovalDescriptor"/>.
    /// </summary>
    public sealed class ApprovalUIHandler :
        StageTypeUIHandlerBase
    {
        #region Fields

        private readonly IViewService viewService;

        private CardRow? settings;
        private IControlViewModel? returnIfNotApprovedFlagControl;
        private IControlViewModel? returnAfterApprovalFlagControl;

        #endregion

        #region Constructors

        /// <summary>
        /// Инициализирует новый экземпляр класса.
        /// </summary>
        /// <param name="viewService"><inheritdoc cref="IViewService" path="/summary"/></param>
        public ApprovalUIHandler(IViewService viewService) =>
            this.viewService = NotNullOrThrow(viewService);

        #endregion

        #region Base Overrides

        /// <inheritdoc/>
        public override Task Initialize(IKrStageTypeUIHandlerContext context)
        {
            if (context.SettingsForms.FirstOrDefault(static i => i.Name == DefaultCardTypes.KrApprovalStageTypeSettingsTypeName) is { } form
                && form.Blocks.FirstOrDefault(static i => i.Name == "ApprovalStageFlags") is { } flagsBlock
                && flagsBlock.Controls.FirstOrDefault(static i => i.Name == "FlagsTabs") is TabControlViewModel flagsTabsViewModel)
            {
                this.returnIfNotApprovedFlagControl = flagsTabsViewModel
                    .Tabs.FirstOrDefault(static i => i.Name == "CommonSettings")
                    ?.Blocks.FirstOrDefault(static i => i.Name == "StageFlags")
                    ?.Controls.FirstOrDefault(static i => i.Name == KrConstants.Ui.ReturnIfNotApproved);

                this.returnAfterApprovalFlagControl = flagsTabsViewModel
                    .Tabs.FirstOrDefault(static i => i.Name == "AdditionalSettings")
                    ?.Blocks.FirstOrDefault(static i => i.Name == "StageFlags")
                    ?.Controls.FirstOrDefault(static i => i.Name == KrConstants.Ui.ReturnAfterApproval);
            }

            this.settings = context.Row;
            this.settings.FieldChanged += this.OnSettingsFieldChanged;

            this.AdvisoryConfigureFields(this.settings.TryGet<bool>(KrConstants.KrApprovalSettingsVirtual.Advisory));
            this.NotReturnEditConfigureFields(this.settings.TryGet<bool>(KrConstants.KrApprovalSettingsVirtual.NotReturnEdit));

            return Task.CompletedTask;
        }

        /// <inheritdoc/>
        public override Task Finalize(IKrStageTypeUIHandlerContext context)
        {
            if (this.settings is not null)
            {
                this.settings.FieldChanged -= this.OnSettingsFieldChanged;
                this.settings = null;
            }

            return Task.CompletedTask;
        }

        #endregion

        #region Private Methods

        private void OnSettingsFieldChanged(object? sender, CardFieldChangedEventArgs e)
        {
            if (e.FieldName == KrConstants.KrApprovalSettingsVirtual.Advisory
                && e.FieldValue is bool advisory)
            {
                this.AdvisoryConfigureFields(advisory);

                if (advisory)
                {
                    if (!this.settings!.Fields.TryGet<Guid?>(KrConstants.KrTaskKindSettingsVirtual.KindID).HasValue)
                    {
                        var (kindID, kindCaption) = this.GetKindAsync(KrConstants.AdvisoryTaskKindID).GetAwaiter().GetResult();
                        if (kindID != Guid.Empty)
                        {
                            this.settings.Fields[KrConstants.KrTaskKindSettingsVirtual.KindID] = kindID;
                            this.settings.Fields[KrConstants.KrTaskKindSettingsVirtual.KindCaption] = kindCaption;
                        }
                    }
                }
                else
                {
                    if (this.settings!.Fields.TryGet<Guid?>(KrConstants.KrTaskKindSettingsVirtual.KindID) ==
                        KrConstants.AdvisoryTaskKindID)
                    {
                        this.settings.Fields[KrConstants.KrTaskKindSettingsVirtual.KindID] = null;
                        this.settings.Fields[KrConstants.KrTaskKindSettingsVirtual.KindCaption] = null;
                    }

                    if (this.returnIfNotApprovedFlagControl is not null)
                    {
                        this.returnIfNotApprovedFlagControl.IsReadOnly = false;
                    }
                }

                return;
            }

            if (e.FieldName == KrConstants.KrApprovalSettingsVirtual.NotReturnEdit
                && e.FieldValue is bool notReturnEdit)
            {
                this.NotReturnEditConfigureFields(notReturnEdit);
            }
        }

        private async Task<(Guid, string)> GetKindAsync(Guid id, CancellationToken cancellationToken = default)
        {
            var taskKindsView = await this.viewService.GetByNameAsync(KrConstants.Views.TaskKinds, cancellationToken);

            var taskKindsViewMetadata = await taskKindsView.GetMetadataAsync(cancellationToken);
            var request = new TessaViewRequest(taskKindsViewMetadata);

            var idParam = new RequestParameterBuilder()
                .WithMetadata(taskKindsViewMetadata.Parameters.FindByName("ID"))
                .AddCriteria(new EqualsCriteriaOperator(), string.Empty, id)
                .AsRequestParameter();

            request.Values.Add(idParam);

            var result = await taskKindsView.GetDataAsync(request, cancellationToken).ConfigureAwait(false);

            if (result.Rows is null
                || result.Rows.Count == 0)
            {
                return default;
            }

            var row = (IList<object>) result.Rows[0];
            return ((Guid) row[0], (string) row[1]);
        }

        /// <summary>
        /// Задаёт настройки полям в соответствии со значением флага "Рекомендательное согласование".
        /// </summary>
        /// <param name="isAdvisory">Значение флага "Рекомендательное согласование".</param>
        private void AdvisoryConfigureFields(bool isAdvisory)
        {
            if (isAdvisory)
            {
                if (this.returnIfNotApprovedFlagControl is not null)
                {
                    this.returnIfNotApprovedFlagControl.IsReadOnly = true;
                    this.settings!.Fields[KrConstants.KrApprovalSettingsVirtual.ReturnWhenDisapproved] = BooleanBoxes.False;
                }
            }
        }

        /// <summary>
        /// Настраивает поля в зависимости от значения флага "Не возвращать на доработку".
        /// </summary>
        /// <param name="isNotReturnEdit">Значение флага <see cref=" KrConstants.KrApprovalSettingsVirtual.NotReturnEdit"/>.</param>
        private void NotReturnEditConfigureFields(bool isNotReturnEdit)
        {
            if (isNotReturnEdit)
            {
                if (this.returnIfNotApprovedFlagControl is not null)
                {
                    this.returnIfNotApprovedFlagControl.ControlVisibility = Visibility.Collapsed;
                }

                if (this.returnAfterApprovalFlagControl is not null)
                {
                    this.returnAfterApprovalFlagControl.ControlVisibility = Visibility.Collapsed;
                }
            }
            else
            {
                if (this.returnIfNotApprovedFlagControl is not null)
                {
                    this.returnIfNotApprovedFlagControl.ControlVisibility = Visibility.Visible;
                }

                if (this.returnAfterApprovalFlagControl is not null)
                {
                    this.returnAfterApprovalFlagControl.ControlVisibility = Visibility.Visible;
                }
            }
        }

        #endregion
    }
}
